core/store/src/parallel_iter.rs

```diff
@@ -0,0 +1,692 @@
+    work_sender: Arc<flume::Sender<WorkQueueItem>>,
+}
+
+impl StoreParallelIterator {
+    /// Creates a new parallel iterator that performs the given task.
+    fn new(options: ParallelIterationOptions, initial_task: WorkItem) -> Self {
+        let (tx, rx) = flume::unbounded::<WorkQueueItem>();
+        let mut threads = Vec::new();
+        let shared_state = Arc::new(ParallelIterationSharedState {
+            num_threads_free: AtomicUsize::new(options.num_threads),
+            thread_work: Mutex::new(vec![None; options.num_threads]),
+        });
+        let tx = Arc::new(tx);
+        tx.send(WorkQueueItem { work: initial_task, work_sender: tx.clone() }).unwrap();
+        drop(tx);
+        for idx in 0..options.num_threads {
+            let rx = rx.clone();
+            let shared_state = shared_state.clone();
+            let options = options.clone();
+            let thread = std::thread::spawn(move || loop {
+                let Ok(WorkQueueItem { work, work_sender }) = rx.recv() else {
+                    // If receiving fails, it means all senders are dropped,
+                    // i.e. there is no more work to do.
+                    break;
+                };
+                shared_state.num_threads_free.fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                shared_state.thread_work.lock().unwrap()[idx] = Some(work.clone());
+
+                // Spend some uninterrupted time to work on the task before checking whether we
+                // should split. This is to avoid unnecessary thrashing when the remaining work
+                // is small. Also, randomize this uninterrupted time, so that when there are free
+                // threads, the other threads don't all try to split at the same time.
+                let min_time_to_split = Instant::now()
+                    + Duration::seconds_f64(
+                        options.min_time_before_split.as_seconds_f64()
+                            * (rand::random::<f64>() + 1.0),
+                    );
+
+                match work {
+                    WorkItem::Range(range) => {
+                        let start = if range.start.is_empty() {
+                            None
+                        } else {
+                            Some(range.start.as_slice())
+                        };
+                        let end =
+                            if range.end.is_empty() { None } else { Some(range.end.as_slice()) };
+                        let iter = options.store.iter_range_raw_bytes(options.column, start, end);
+
+                        for kv in iter {
+                            let (key, value) = kv.unwrap();
+                            (options.callback)(&key, Some(&value));
+                            if Instant::now() > min_time_to_split {
+                                if shared_state
+                                    .num_threads_free
+                                    .load(std::sync::atomic::Ordering::Relaxed)
```

Comment:
since it's just Relaxed, if you read it and it's greater than 0, i think theres no guarantee that theres at least 1 thread that is not currently in the loop doing some work right? i bet it's not a big deal in practice esp on x86, and also since there's the `shared_state.thread_work.lock()` right after setting it, which is prob enough of a barrier. But in any case `Relaxed` feels a little fragile to me for what's being done here

Reply:
There's no guarantee about anything in num_threads_free. It doesn't matter anyway, it's just a signal to suggest the current workers that it's probably wise to split. Split or not, it doesn't affect correctness. Thus the use of Relaxed.